### DISCUSSION TITLE
Libmesh update and rebuild error

### DISCUSSION THREAD

**Discussion by Balbest** [idaholab/moose#27559] [Q&A General] [answered]
I got the following error when run update_and_rebuild_libmesh.sh:  DEBUG and NDEBUG should never be defined simultaneously.
I don't know if it's relevant, but I also unset LIBMESH_DIR to start the update beforehand

**Comment by milljm** [comment]
Do you have `METHODS` set in your environment?

Can you run the diagnostics.sh script in the same directory, and report those results:
```bash
cd moose/scripts
./diagnostics.sh
```

**Reply by Balbest** [reply]
No METHODS set.

Here are the results of diagnostics.sh:
<details><summary>diagnostics</summary>
<p>
##################################################################################################
Influential Environment Variables

CC_FOR_BUILD=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-cc
CC=mpicc
CFLAGS=-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/balbest/miniforge/envs/moose/include
CONDA_BUILD_SYSROOT=/home/balbest/miniforge/envs/moose/x86_64-conda-linux-gnu/sysroot
CONDA_CHANNEL=https://conda.software.inl.gov/public
CONDA_DEFAULT_ENV=moose
CONDA_EXE=/home/balbest/miniforge/bin/conda
CONDA_PREFIX_1=/home/balbest/miniforge
CONDA_PREFIX=/home/balbest/miniforge/envs/moose
CONDA_PROMPT_MODIFIER=(moose) 
CONDA_PYTHON_EXE=/home/balbest/miniforge/bin/python
CONDA_SHLVL=2
CONDA_TOOLCHAIN_BUILD=x86_64-conda-linux-gnu
CONDA_TOOLCHAIN_HOST=x86_64-conda-linux-gnu
CPPFLAGS=-DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/balbest/miniforge/envs/moose/include
CPP=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-cpp
CURL_CA_BUNDLE=
CXXFILT=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-c++filt
CXXFLAGS=-fvisibility-inlines-hidden  -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/balbest/miniforge/envs/moose/include -std=c++17
CXX_FOR_BUILD=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-c++
CXX=mpicxx
F77=mpif77
F90=mpif90
F95=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-f95
FC_FOR_BUILD=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-gfortran
FC=mpif90
FFLAGS=-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/balbest/miniforge/envs/moose/include
HDF5_DIR=/home/balbest/miniforge/envs/moose
LDFLAGS=-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/balbest/miniforge/envs/moose/lib -Wl,-rpath-link,/home/balbest/miniforge/envs/moose/lib -L/home/balbest/miniforge/envs/moose/lib
LD_GOLD=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-ld.gold
LD=/home/balbest/miniforge/envs/moose/bin/x86_64-conda-linux-gnu-ld
MOOSE_JOBS=6
MOOSE_NO_CODESIGN=true
PATH=/home/balbest/miniforge/envs/moose/bin:/home/balbest/miniforge/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/balbest/miniforge/envs/moose/wasp/bin
PETSC_DIR=/home/balbest/miniforge/envs/moose
REQUESTS_CA_BUNDLE=
SSL_CERT_FILE=
WASP_DIR=/home/balbest/miniforge/envs/moose/wasp

##################################################################################################
Compiler(s) (CC CXX FC F77 F90):

CC=/home/balbest/miniforge/envs/moose/bin/mpicc
CC -show:
x86_64-conda-linux-gnu-cc -I/home/balbest/miniforge/envs/moose/include -I/home/balbest/miniforge/envs/moose/include -L/home/balbest/miniforge/envs/moose/lib -Wl,-rpath,/home/balbest/miniforge/envs/moose/lib -I/home/balbest/miniforge/envs/moose/include -L/home/balbest/miniforge/envs/moose/lib -Wl,-rpath -Wl,/home/balbest/miniforge/envs/moose/lib -Wl,--enable-new-dtags -lmpi
CC version:	x86_64-conda-linux-gnu-cc (conda-forge gcc 10.4.0-19) 10.4.0

CXX=/home/balbest/miniforge/envs/moose/bin/mpicxx
CXX -show:
x86_64-conda-linux-gnu-c++ -I/home/balbest/miniforge/envs/moose/include -I/home/balbest/miniforge/envs/moose/include -L/home/balbest/miniforge/envs/moose/lib -Wl,-rpath,/home/balbest/miniforge/envs/moose/lib -I/home/balbest/miniforge/envs/moose/include -L/home/balbest/miniforge/envs/moose/lib -lmpicxx -Wl,-rpath -Wl,/home/balbest/miniforge/envs/moose/lib -Wl,--enable-new-dtags -lmpi
CXX version:	x86_64-conda-linux-gnu-c++ (conda-forge gcc 10.4.0-19) 10.4.0

FC=/home/balbest/miniforge/envs/moose/bin/mpif90
FC -show:
x86_64-conda-linux-gnu-gfortran -I/home/balbest/miniforge/envs/moose/include -fallow-argument-mismatch -L/home/balbest/miniforge/envs/moose/lib -Wl,-rpath,/home/balbest/miniforge/envs/moose/lib -I/home/balbest/miniforge/envs/moose/include -I/home/balbest/miniforge/envs/moose/include -L/home/balbest/miniforge/envs/moose/lib -lmpifort -Wl,-rpath -Wl,/home/balbest/miniforge/envs/moose/lib -Wl,--enable-new-dtags -lmpi
FC version:	GNU Fortran (conda-forge gcc 10.4.0-19) 10.4.0

F77=/home/balbest/miniforge/envs/moose/bin/mpif77
F77 -show:
x86_64-conda-linux-gnu-gfortran -I/home/balbest/miniforge/envs/moose/include -fallow-argument-mismatch -L/home/balbest/miniforge/envs/moose/lib -Wl,-rpath,/home/balbest/miniforge/envs/moose/lib -I/home/balbest/miniforge/envs/moose/include -I/home/balbest/miniforge/envs/moose/include -L/home/balbest/miniforge/envs/moose/lib -lmpifort -Wl,-rpath -Wl,/home/balbest/miniforge/envs/moose/lib -Wl,--enable-new-dtags -lmpi
F77 version:	GNU Fortran (conda-forge gcc 10.4.0-19) 10.4.0

F90=/home/balbest/miniforge/envs/moose/bin/mpif90
F90 -show:
x86_64-conda-linux-gnu-gfortran -I/home/balbest/miniforge/envs/moose/include -fallow-argument-mismatch -L/home/balbest/miniforge/envs/moose/lib -Wl,-rpath,/home/balbest/miniforge/envs/moose/lib -I/home/balbest/miniforge/envs/moose/include -I/home/balbest/miniforge/envs/moose/include -L/home/balbest/miniforge/envs/moose/lib -lmpifort -Wl,-rpath -Wl,/home/balbest/miniforge/envs/moose/lib -Wl,--enable-new-dtags -lmpi
F90 version:	GNU Fortran (conda-forge gcc 10.4.0-19) 10.4.0

OK

##################################################################################################
Python Sanity Checks

Verify `/usr/bin/env python3 --version` (reporting as: Python 3.10.8),
matches versions for: `which python3 && which python`

OK

##################################################################################################
Python Modules (TestHarness, run-ability)

OK

##################################################################################################
CONDA MOOSE Packages

moose-dev                 2024.05.02              build_0    https://conda.software.inl.gov/public
moose-libmesh             2024.04.03              build_1    https://conda.software.inl.gov/public
moose-libmesh-vtk         9.2.6                   build_9    https://conda.software.inl.gov/public
moose-mpich               4.0.2                  build_16    https://conda.software.inl.gov/public
moose-peacock             2023.04.11           hb6770a3_0    https://conda.software.inl.gov/public
moose-petsc               3.20.3                  build_1    https://conda.software.inl.gov/public
moose-tools               2024.05.02           h4a78fc2_0    https://conda.software.inl.gov/public
moose-wasp                2024.04.23              build_0    https://conda.software.inl.gov/public

OK Conda packages and MOOSE repo/submodule versions match

</p>
</details>

**Reply by GiudGiud** [reply, marked as answer]
did you change the CPPFLAGS environment variable to include this flag? or maybe something in your environment did.
```
CPPFLAGS=-DNDEBUG
```
we would add -DDEBUG when building the debug version, which creates the problem

maybe let's edit this `DNDEBUG` away from CPPFLAGS

**Reply by milljm** [reply]
Because you have the top most level package `moose-dev` installed, there is no need to build libMesh manually. The `moose-dev` Conda package will provide all the necessary libraries.

If you are wanting to build libMesh yourself we should create a separate environment specially tailored to do so. That way you don't have to worry about setting or un-setting LIBMESH_DIR, etc.
Of course, this is neither here nor there for the error you are receiving... Just my observations!

**Reply by Balbest** [reply]
Thank you GuidGuid and milljm! The reason I did it is the problem with peacock, and this was one of the solutions I found on forums. It didn't help, as one could expect. So I'll create another topic about this.

I didn't change CPPFLAGS intentionally. Moreover, the installations of both conda and moose are absolutely fresh.
